### PR TITLE
feat(machines): select machines using the new types

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Update the checkboxes in the machine list to select filters, groups and machines.

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to the machine list.
- Tick some machines.
- Open the redux dev tools and look at `machine` -> `selectedMachines` and you should see the machine ids.
- Tick a group and you should see the group in the selected state (not all grouping is currently working so you might have to change the table grouping to find one that the API supports - owner was working for me).
- Tick the checkbox in the table header (next to FQDN) and the selected state should show the filter (probably empty unless you've set one).

## Fixes

Fixes: canonical/app-tribe#1126.
Fixes: canonical/app-tribe#1257.

### CODE REVIEW

**Comment by huwshimi** [comment]
I've split out a number of smaller PRs so this will be on hold until those land.

**Comment by huwshimi** [comment]
This is ready for review now.

File: src/app/machines/views/MachineList/MachineListTable/AllCheckbox/AllCheckbox.test.tsx
Comment (petermakowski) [inline src/app/machines/views/MachineList/MachineListTable/AllCheckbox/AllCheckbox.test.tsx:34]: Might be worth testing in combination with a label (e.g. `expect(screen.getByRole("checkbox", { name: /all/ })).not.toBeChecked();`). This checkbox currently has none:

image

File: src/app/machines/views/MachineList/MachineListTable/AllCheckbox/AllCheckbox.test.tsx
Comment (petermakowski) [inline src/app/machines/views/MachineList/MachineListTable/AllCheckbox/AllCheckbox.test.tsx:52]: Somebody's embracing custom jest matchers 👌

File: src/app/machines/components/TableCheckbox/TableCheckbox.tsx
Comment (petermakowski) [inline src/app/machines/components/TableCheckbox/TableCheckbox.tsx:23]: What do you think about making the `inputLabel` required for this component?

Looks like - in contrary to the general purpose `Input` from `@canonical/react-components` - this component isn't meant to be wrapped in a `label` element which would provide it.

File: src/app/machines/views/MachineList/MachineListTable/GroupCheckbox/GroupCheckbox.tsx
Comment (petermakowski) [inline src/app/machines/views/MachineList/MachineListTable/GroupCheckbox/GroupCheckbox.tsx:68]: I was initially a bit confused why is this done inside of a component handler function. It would be more consistent with the rest of the codebase if this piece of logic was handled in a redux reducer.

**Review by petermakowski — APPROVED** [review]
I left some comments, but feel free to address separately if you prefer.

**Comment by petermakowski** [comment]
Hopefully this won't be too confusing for users that you can no longer de-select individual machines/groups of machines once you selected a group/all machines on a page.

File: src/app/machines/views/MachineList/MachineListTable/AllCheckbox/AllCheckbox.test.tsx
Comment (huwshimi) [reply src/app/machines/views/MachineList/MachineListTable/AllCheckbox/AllCheckbox.test.tsx:34]: Done!

File: src/app/machines/components/TableCheckbox/TableCheckbox.tsx
Comment (huwshimi) [reply src/app/machines/components/TableCheckbox/TableCheckbox.tsx:23]: I think that would work if it wasn't for `AllCheckbox` which just has an `aria-label` (or at least it will do now that you've pointed out that it's missing :)).

File: src/app/machines/views/MachineList/MachineListTable/GroupCheckbox/GroupCheckbox.tsx
Comment (huwshimi) [reply src/app/machines/views/MachineList/MachineListTable/GroupCheckbox/GroupCheckbox.tsx:68]: I've moved this part into a new issue just because I had a couple of questions: https://github.com/canonical/maas-ui/issues/4375.

**Pull request merged**
